File: apps/backend/app/contracts/learner.py
"""Learner: LearnerState, Misc, LearnerResponse (Architecture Document §6.7, §6.8).

LearnerState is the student's mental model, updated each turn. It is never
shared with the Evaluator as an answer key (invariant §1.4). LearnerResponse
is always in the student role — asking, doubting, or paraphrasing.
"""

from __future__ import annotations

from pydantic import Field

from .base import CamelModel
from .enums import DerivedFrom, LearnerResponseType


class Misc(CamelModel):
    """A single active misconception the student currently holds (§6.7)."""

    concept: str = Field(description="The misunderstood concept")
    belief: str = Field(description="The student's faulty belief about it")


class LearnerState(CamelModel):
    """The student's mental model, updated each turn (§6.7)."""

    session_id: str = Field(description="Session that owns the state")
    understood_concepts: list[str] = Field(
        default_factory=list, description="Concepts the student now understands"
    )
    active_misconceptions: list[Misc] = Field(
        default_factory=list, description="Active misconceptions { concept, belief }"
    )
    open_gaps: list[str] = Field(
        default_factory=list, description="Understanding gaps not yet filled"
    )
    questions_asked: list[str] = Field(
        default_factory=list, description="Questions already asked"
    )
    updated_at_turn: int = Field(
        default=0, description="Last turn that updated the state"
    )


class LearnerResponse(CamelModel):
    """The Learner's output for one turn — always in the student role (§6.8)."""

    response_id: str = Field(description="Unique response identifier")
    turn_index: int = Field(description="Turn the response came from")
    type: LearnerResponseType = Field(
        description='"question" | "confusion" | "acknowledgment" | "paraphrase"'
    )
    text: str = Field(description="The student's utterance")
    target_concept: str | None = Field(
        default=None, description="Concept the question targets"
    )
    derived_from: DerivedFrom = Field(
        description='"gap" | "misconception" | "new_info"'
    )

File: apps/backend/src/contracts/learner.ts
/**
 * Learner: LearnerState, Misc, LearnerResponse (Architecture Document §6.7, §6.8).
 *
 * LearnerState is the student's mental model, updated each turn. It is never
 * shared with the Evaluator as an answer key (invariant §1.4). LearnerResponse
 * is always in the student role — asking, doubting, or paraphrasing.
 */

import { z } from "zod";
import { derivedFromSchema, learnerResponseTypeSchema } from "./enums.js";

/** A single active misconception the student currently holds (§6.7). */
export const miscSchema = z.object({
  concept: z.string(),
  belief: z.string(),
});
export type Misc = z.infer<typeof miscSchema>;

/**
 * How often the student has already asked about one core concept. Two is the
 * limit, after which it accepts the explanation and asks to move on
 * (agents/learner/learner.repeat.ts). Optional: states persisted before the
 * limit existed carry no tally.
 */
export const askedConceptSchema = z.object({
  key: z.string(),
  label: z.string(),
  count: z.number().int(),
  /**
   * "probe" (the default) counts plain questions about the concept; "extend"
   * counts questions that apply it to a new case, which get their own budget.
   */
  kind: z.enum(["probe", "extend"]).optional(),
});
export type AskedConcept = z.infer<typeof askedConceptSchema>;

/** The student's mental model, updated each turn (§6.7). */
export const learnerStateSchema = z.object({
  sessionId: z.string(),
  understoodConcepts: z.array(z.string()).default([]),
  activeMisconceptions: z.array(miscSchema).default([]),
  openGaps: z.array(z.string()).default([]),
  questionsAsked: z.array(z.string()).default([]),
  askedConcepts: z.array(askedConceptSchema).optional(),
  /**
   * Questions in a row that asked how the teacher's own previous answer works
   * (agents/learner/learner.depth.ts). Optional: older states carry none.
   */
  followUpDepth: z.number().int().optional(),
  updatedAtTurn: z.number().int().default(0),
});
export type LearnerState = z.infer<typeof learnerStateSchema>;

/** The Learner's output for one turn — always in the student role (§6.8). */
export const learnerResponseSchema = z.object({
  responseId: z.string(),
  turnIndex: z.number().int(),
  type: learnerResponseTypeSchema,
  text: z.string(),
  targetConcept: z.string().nullable().optional(),
  derivedFrom: derivedFromSchema,
});
export type LearnerResponse = z.infer<typeof learnerResponseSchema>;
